mne/preprocessing/ica.py

```diff
@@ -682,68 +692,50 @@ def _fit(self, data, fit_type):
-        # ensure that all internals of the _PCA instance are properly updated
-        # as well, and we don't need to handle this case specially below.
-        # Note that the result should be numerically identical to the first
-        # PCA run for all retained components.
-        if (isinstance(self.max_pca_components, float) and
-                self.max_pca_components != 1.0):
-            del data_transformed  # Free memory.
-            self.max_pca_components_, _ = _exp_var_ncomp(
-                pca.explained_variance_ratio_, self.max_pca_components)
-
-            if (self.n_components is not None and
-                    self.max_pca_components_ < self.n_components):
-                raise ValueError(
-                    f'You asked to select only a subset of PCA components by '
-                    f'passing max_pca_components={self.max_pca_components}, '
-                    f'which equals {self.max_pca_components_} components '
-                    f'for this specific dataset. However, you also '
-                    f'requested to pass {self.n_components} of those '
-                    f'components to ICA, which is mathematically '
-                    f'impossible. Please either increase '
-                    f'max_pca_components, set it to None, reduce '
-                    f'n_components, or set n_components=None')
-
-            pca = _PCA(n_components=self.max_pca_components_, whiten=True)
-            data_transformed = pca.fit_transform(data.T)
-
-        assert data_transformed.shape == (n_samples, self.max_pca_components_)
-        del data
+        pca = _PCA(n_components=self._max_pca_components, whiten=True)
+        data = pca.fit_transform(data.T)
+        use_ev = pca.explained_variance_ratio_
+        n_pca = self.n_pca_components
+        if isinstance(n_pca, float):
+            n_pca = int(_exp_var_ncomp(use_ev, n_pca)[0])
+        elif n_pca is None:
+            n_pca = len(use_ev)
+        assert isinstance(n_pca, (int, np.int_))
 
         # If user passed a float, select the PCA components explaining the
         # given cumulative variance. This information will later be used to
         # only submit the corresponding parts of the data to ICA.
-        if isinstance(self.n_components, float):
+        msg = None
+        if self.n_components is not None:
+            self.n_components_ = self.n_components
+        elif self.n_pca_components is not None:
+            self.n_components_ = self.n_pca_components
+        else:  # Both None
+            self.n_components_ = len(pca.components_)
+            msg = 'Selecting all PCA components'
+        if isinstance(self.n_components_, float):
+            # XXX not 100% sure if this should be use_ev or use_ev[:n_pca],
+            # i.e., is n_components float the threshold for the explained
+            # variance remaining after subselecting n_pca_components, or is
+            # it from the original data
```

Comment:
?

Reply:
This is the point at the top of the issue that needs resolving.

Concete example. On `master`
- Assumue data is 4 channels with PCA exp var ratios 0.4, 0.3, 0.2, 0.1
- Doing `max_pca_components=3, n_components=1.` (or None) will yield only 3 components to ICA (`n_components_=3`)
- Doing `max_pca_components=4, n_components=1.` (or None) will yield 4 components in the ICA (`n_components_=4`).

This is because the `n_components=1.` is relative *to the variance explained by the max_pca_components*, not of the entire data (if it were the entire data, the first case above would need to raise an error because you'd be requesting 4 ICA components but only using 3 PCA ones).

If we kill `max_pca_components`, what should `n_pca_components=3, n_components=1.` mean? Should it yield `n_components_=4`, or `n_components_=3`? In other words, should the float `n_components` be relative to:

1. The explained variance across just `n_pca_components` -- this is my preference; it's implemented here as it required fewer tweaks to the tests; and is consistent with the behavior before if we're thinking of the `n_pca_components` in init now taking over the role of `max_pca_components`
2. The total explained variance in your data?

The only problem I've seen with (1) is that it means that code that looked like:

- `max_pca_components=None, n_pca_components=3, n_components=1.`

in master will have `n_components_=4` and on this PR it will have `n_components_=3`. But everywhere in our code base where we set `n_pca_components=N` as an int we had `max_pca_components=N` as well, so I'm hoping this might qualify as a bug fix of sorts (it's weird to have more components in the ICA step than you plan on reconstructing in the PCA step).